### PR TITLE
ColorHsv addition

### PR DESCRIPTION
**Pull request by StatisMike**
Introduced representation of HSV color model on the Rust side

Currently, there is a possibility of creating `Color` GodotType from HSV values with `Color::from_hsv()`, but there is no option to reverse the process (receive Hue, Saturation, and Value values out of `Color`), which makes it problematic to fully use `HSV` representation.

This PR introduces `ColorHsv` Rust type, which can be converted from and into `Color` GodotType, resolving the issue. 

## Conversion algorithms

Two different algorithm for conversion were found during research:
- Smith, Alvy Ray. "Color gamut transform pairs." ACM Siggraph Computer Graphics 12.3 (1978): 12-19
  - uses float RGB and HSV values (and Color in Godot uses float RGB under the hood, its `r8()`, `g8()` and `b8()` getters are  clamping the floats into 0-255 integer range
  -  is widely used
- Chernov, Vladimir, Jarmo Alander, and Vladimir Bochko. "Integer-based accurate conversion between RGB and HSV color spaces." Computers & Electrical Engineering 46 (2015): 328-337.
  - uses integer RGB and HSV values
  - is slightly more tricky to implement
  - by paper authors claim it should be faster than float-based one

## Benchmarks and chosen algorithm

During development, both algorithms were implemented, but (at least for now) the float-based was chosen. Integer-based could be potentially implemented in the future if such a need arises. Rationale based on debug-build benchmarks:

```
   color.rs:
   -- godot_from_hsv             ...      0.248μs      0.372μs
   -- rust_from_hsv              ...      0.055μs      0.056μs
   -- rgb_to_hsv_roundtrip       ...      0.086μs      0.086μs
   -- rgb_to_hsv_mutate_roundtri ...      0.098μs      0.124μs
```

The float-based algorithm is already very fast - as we need an FFI call from Rust to Godot to pass the `Color` itself, further improvements could be negligible. Furthermore, it's much faster (possibly partly because of the above) than Godot's `HSV->RGB` already.
- `godot_from_hsv` - HSV -> RGB via `Color::from_hsv()` 
- `rust_from_hsv` - HSV -> RGB via proposed implementation, 
- `rgb_to_hsv_roundtrip` - RGB -> HSV -> RGB 
- `rgb_to_hsv_mutate_roundtrip` - same as above, but HSV values are also mutated in between

Additionally, during development, the integer-based algorithm was several times slower than the float-based one, though it is possible that my implementation was faulty in some way (as it is a little trickier to implement), or just the first retrieval of integer RGB values from `Color` and their later conversion from integers to floats could be partly at blame.

An additional benefit is that by keeping the `ColorHsv` representation as `f32` based, it is uniform with the main `Color` representation. 

Some tradeoffs are mentioned in the docs (and the tests are put in place to make sure it is stable):
- the float-based algorithm is not entirely lossless - while `Color::from_hsv()` and implemented `ColorHsv::into<Color>()` results are equal, multiple repeatable conversions between the formats can introduce some loss (less than 0.001 in every value).
- the integer-based algorithm has the potential to be more stable (as it operates fully on integers), but it isn't a guarantee (as the integers needs to be converted to `f32` to construct `Color`, and then retrieved with `Color.*8()` getters.

## Possible refactor
The first commit introduces the logic - after it is approved, I would aim to also refactor the `color` module with the second commit. Planned changes would be 
- at minimum moving it to another directory along with `hsv.rs` (renamed `color_hsv.rs`) and `color_constants.rs`
- at maximum additionally unpacking some code from `color.rs` to separate modules could be also beneficial, as the file is already pretty verbose.

An additional refactor would be to move `ColorHsv` from `godot::builtins` and `godot::prelude`. Maybe make the `godot::builtins::color` namespace? Or move `ColorHsv` from `godot::builtins` entirely, maybe to `godot::engine::colors` (as it isn't technically a `GodotType`, but a helper struct).

Update:
Benchmarks from `itest` CI execution on linux shows a slightly different picture than above (though Godot's `Color::from_hsv()` is still slower):

```
                                              min       median
   color.rs:
   -- godot_from_hsv             ...      0.190μs      0.192μs
   -- rust_from_hsv              ...      0.042μs      0.042μs
   -- rgb_to_hsv_roundtrip       ...      0.080μs      0.081μs
   -- rgb_to_hsv_mutate_roundtri ...      0.095μs      0.096μs
```

### CODE REVIEW

**Comment by GodotRust** [comment]
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-605

File: godot-core/src/builtin/color_hsv.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        }
+    }
+
+    /// Construct from Hue, Saturation, Value where Hue is represented in deegrees (`0.0` - `360.0`).
+    ///
+    /// Saturation, Value and Alpha have to be provided in range between `0.0` an `1.0`. If values outside of valid ranges are provided,
+    /// they will be clamped to a valid range. To add `alpha` value, use `with_alpha()` method.
+    pub fn from_hsv_deg(h: f32, s: f32, v: f32) -> Self {
+        ColorHsv {
+            h: h.clamp(0., 360.) / 360.,
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:109]: I'm not sure if `clamp` is the most intuitive behavior here -- if you have an angle, wouldn't you rather wrap it? (i.e. modulo 360°)

File: godot-core/src/builtin/color_hsv.rs [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+            v: v.clamp(0., 1.),
+            a: 1.,
+        }
+    }
+
+    /// Adds an alpha value.
+    pub fn with_alpha(mut self, alpha: f32) -> Self {
+        self.a = alpha.clamp(0., 1.);
+        self
+    }
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:120]: Not too much a fan of selective builders 😉 I'm also not sure if we should keep `Color::with_alpha`.

I understand the purpose to enhance constructors, but in this case, it's probably OK to have 3 constructors:
```
from_hsv
from_hsva
from_hsva_deg
```
(the degree one is rather uncommon, so specifying 1.0 for `a` should be acceptible)

File: godot-core/src/builtin/color_hsv.rs [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        self.v = v.clamp(0., 1.);
+    }
+
+    pub fn alpha(&self) -> f32 {
+        self.a
+    }
+
+    pub fn set_alpha(&mut self, alpha: f32) {
+        self.a = alpha.clamp(0., 1.);
+    }
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:160]: I assume the main idea of this encapsulation is to ensure that values always stay in bounds.

While I agree that this improves invariants, most other builtin types do not have an API that follows such a design (and it's typically not enforcable because Godot can modify those). An alternative approach would also be to have assertions only in places that _use_ those invariants (e.g. conversions).

Some more input on this would be welcome 🙂

File: godot-core/src/builtin/color_hsv.rs [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+impl From<ColorHsv> for Color {
+    fn from(value: ColorHsv) -> Self {
+        let (r, g, b, a) = hsv_to_rgb_f(value.h, value.s, value.v, value.a);
+
+        Color { r, g, b, a }
+    }
+}
+
+fn rgb_to_hsv_f(r: f32, g: f32, b: f32, a: f32) -> (f32, f32, f32, f32) {
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:179]: Probably `rgba_to_hsva` is more descriptive.

Why the `f`? float? Is it really needed, as there is no other method with a different representation?

File: godot-core/src/builtin/color_hsv.rs [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+    if s < 0.0001 {
+        return (v, v, v, a);
+    }
+
+    let h = h * 6.;
+
+    let i = h.floor();
+
+    let f = h - i;
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:227]: Could be grouped:
```suggestion
    let h = h * 6.;
    let i = h.floor();
    let f = h - i;
```

**Review by Bromeon — COMMENTED** [review]
Thank you, this looks very polished! 👍 

In general, for contributions of this scope, I would appreciate some initial design discussion in a GitHub issue. Especially in this case, where the functionality goes slightly beyond what Godot provides, it might be nice to get some more opinions first. In other areas, I was a bit hesitant for RIIR approaches purely for performance reasons (recently quaternions), due to added maintenance and bug risk.

Here, the implementation seems relatively small and well-tested, but I'd still appreciate other opinions on the approach 🙂

Independently, I'll already give some more detailed feedback.

File: godot-core/src/builtin/color_hsv.rs [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+/// HSV floating-number Color representation.
+///
+/// Godot's [`Color`] built-in type supports mainly RGBA floating point-based notation. `ColorHsv` supports manipulating its HSV
+/// representation by introducing conversion methods between itself and [`Color`].
+///
+/// `ColorHsv` *is not* a [`GodotType`](crate::builtin::meta::GodotType). To use it in properties expecting [`Color`], you need to convert
+/// it back to this type.
+///
+/// [`Color`] created by `ColorHsv::into::<Color>()` is equal to one created by `Color::from_hsv()`, but its execution time is approximately
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:18]: It's enough to make only the first occurrence of `Color` a link.

But you could maybe add a link to `ColorHsv` in the `Color` docs 🙂

**Comment by Bromeon** [comment]
Benchmark on Windows 10, using `cargo build --release -p itest` and a Godot `4.3.dev.custom_build.0f8c955c1` release template:
```
                                              min       median
   color.rs:
   -- godot_from_hsv             ...      0.057μs      0.058μs
   -- rust_from_hsv              ...      0.011μs      0.011μs
   -- rgb_to_hsv_roundtrip       ...      0.014μs      0.015μs
   -- rgb_to_hsv_mutate_roundtri ...      0.027μs      0.028μs
```

File: godot-core/src/builtin/color_hsv.rs (hunk 2 above)
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:120]: The *uncommon* could be actually more *common* than it seems to us. All online convertors for RGB<->HSV uses `0..=360` or `0.0..360.0` for hue. Though I guess if you use alpha value then possibly you are using hue `0.0..=1.0`

`with_alpha` is introduced to be analogous with `Color::with_alpha`, if the `Color` one is planned for removal I can remove it there too.

File: godot-core/src/builtin/color_hsv.rs (hunk 4 above)
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:179]: Yeah, it was left from the two different implementations as I've described in the initial comment (`f` and `i`). Will remove, good catch!

File: godot-core/src/builtin/color_hsv.rs (hunk 3 above)
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:160]: Yeah - this is to not allow modifying values beyond the valid range. From what I've understood, the `Color` allows for values beyond that range for a reasons which wouldn't be reasonable with HSV representation. As described in `Color` doc:

> Channel values are typically in the range of 0 to 1, but this is not a requirement, and values outside this range are explicitly allowed for e.g. High Dynamic Range (HDR).

**Comment by StatisMike** [comment]
@Bromeon Thanks for the review, will address the current points soon.

I would like in near future to test the integer convertion algorithm to make sure that its low speed is avoidable OR its results are more stable than float one. Either way I think float-based should be provided, as it is already very fast and Color is represented by floats under the hood in Godot).

I will try to start some discussion on Discord to allow other members of the community to take a look and place their opinions. This was intended to be much smaller in scope but snowballed a little, and many of the caveats came to light during actual implementation... Actually, I think the PR itself is an appropriate place for discussion and I didn't want to prematurely put it to the light for other maintainers as everyone seems to be working on much more important core features.

The thing that I've also though of, especially seeing your comment on clamping values:
>  most other builtin types do not have an API that follows such a design (and it's typically not enforcable because Godot can modify those)

As the `ColorHsv` isn't a `GodotType` itself, maybe it should be actually moved to `godot::engine`?

File: godot-core/src/builtin/color_hsv.rs [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+// Manual implementation because Hue minimal value (0.) is equal to maximal value (1.)
+impl PartialEq for ColorHsv {
+    fn eq(&self, other: &Self) -> bool {
+        (self.h == other.h || (self.h - other.h).abs() == 1.)
+            && self.s == other.s
+            && self.v == other.v
+            && self.a == other.a
+    }
+}
```
[the remark this answers is not in the document]
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:88]: For one point, the hue: 0. and hue 1. results in the same RGB color. Actually, I think that the `PartialOrd` should be removed there (as it doesn't really make much sense to check if one color is greater than other). You can compare the values in the fields themselves.

File: godot-core/src/builtin/color_hsv.rs [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+/// assert_ne!(color1, color2);
+///
+/// assert!(diff_r < 0.001);
+/// assert!(diff_g < 0.001);
+/// assert!(diff_b < 0.001);
+/// ```
+///  
+/// ## Reference
+/// - Smith, Alvy Ray. "Color gamut transform pairs." ACM Siggraph Computer Graphics 12.3 (1978): 12-19.
+#[derive(Debug, Clone, Copy, PartialOrd, Default)]
```
[the remark this answers is not in the document]
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:72]: @Bromeon Good catch, I forgot about alpha channel, it now creates transparent black by default

**Comment by Bromeon** [comment]
> I would like in near future to test the integer convertion algorithm to make sure that its low speed is avoidable OR its results are more stable than float one. Either way I think float-based should be provided, as it is already very fast and Color is represented by floats under the hood in Godot).

I would wait with spending more engineering effort on this front -- there is no evidence that people need a HSV algorithm that is distinct from Godot's _or_ that the improved performance really matters.

Generally I try to keep godot-rust development very pragmatic: we need to have a concrete problem before a solution is justified. Which is also why I'm hesitant on rewrite-it-in-Rust PRs 😉 

---




> As the `ColorHsv` isn't a `GodotType` itself, maybe it should be actually moved to `godot::engine`?

No, `godot::engine` is mostly for generated code (apart from `load` etc, but we need to clean those up at some point). I think `godot::builtin` is fine, there is quite a bit of manual code in there as well. And it's the most comfortable for a user: if they import `ColorHsv`, chances are high that they also need `Color`, at which point both would come from the same module path.

File: godot-core/src/builtin/color_hsv.rs (hunk 1 above)
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:109]: Great suggestion - not only there but for another constructor and both setters also. It would allow easily rotating around hue circle and possibly also remove the issue with hue `PartialEq`

**Comment by Ogeon** [comment]
My 5 cents, as someone who has messed around with colors for a while:

* It makes sense to stick to floats, to avoid rounding to integers. This may be important if the RGB values are treated as linear gamma, since rounding can cause banding in darker shades.
* I agree regarding not clamping the hue. It's an angle/direction, even though it's sometimes represented as prevents (there's all sorts of variants). You will get a more useful behavior if it's wrapped.
* Don't forget about negative hues when wrapping!
* Hue is generally wonky to deal with. I chose to have a wrapper type in Palette, that imposed rules like not having multiplication or division, since those aren't well defined. Ordering is also weird.
* I haven't reviewed the implementation entirely, but watch out for unusual values, like "saturated black", that may or may not result in 0/0 somewhere. Not sure if you covered all the cases already.
* Watch out for the color space rabbit hole. It's deep. 🐇

**Comment by StatisMike** [comment]
@Ogeon Thanks for your input! Some of your points will be adressed below, but where they aren't:

- the integer vs floats I will certainly revisit. I believe that the integer-based algorithm could really allow for stable RGB<->HSV roundtrips, but there are unfortunately some strings attached that are imposed by RGB color being represented in Godot as floats. Maybe I will revisit the topic in the future, but for now I am pretty happy with the performance, ease of use and general usefullness of the float-based one. The integer-based one is more trickier to implement alongside the whole wrapper around it. You can look for the article referenced in original post for details, if it picks your interests :)
- negative hues are also covered, and relevant tests are added :)
- I believe that all of the edge cases like *saturated black* are be covered, the algorithms itself were luckily not concocted by me, just the implementation in Rust was. The only thing that I've found that wasn't really covered in original is `hue == 1.0` situation, possibly because the authors were wrapping the value to not allow this situation (as in practice `(hue == 1.0) == (hue == 0.0) `

@Bromeon Applied the effects of our discord discourse on this topic. Brief summary of the results of it:

- ultimately dropped completely the getters and setters as it allows for less verbose syntax when incrementing/decrementing: `hsv_color.h += 0.1;` vs `hsv_color.set_h(hsv_color.h() + 0.1)` and won't enforce the strategy of wrapping vs clamping for hue. 
- no built-in clamping or wrapping is taking place anywhere. Instead, the conversions would panic if values are outside of valid `0.0..=1.0` range, and useful normalizing methods are introduced: `Color::normalize()` for RGBA clamp and `ColorHsv::normalize_clamped_h()` and `ColorHsv::normalize_wrapped_h()` for HSVA clamp or Hue wrap + SVA clamp.
- *degrees*-accepting constructor for `ColorHsv` was dropped

File: godot-core/src/builtin/color.rs [hunk 9]
```diff
@@ -196,8 +198,6 @@ impl Color {
         self.a = from_u8(a);
     }
 
-    // TODO add getters and setters for h, s, v (needs generated property wrappers)
```
Comment (StatisMike) [inline godot-core/src/builtin/color.rs:199]: I believe the `ColorHsv` addition removes the necessity of this TODO

File: godot-core/src/builtin/color_hsv.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:95]: I've removed the `PartialOrd` derive, as I don't think it actually makes sense logically. As the *automatic wrapping* was dropped, the 0.0 == 1.0 should be IMO upkept for this to be not only mathematically, but also logically correct.

File: godot-core/src/builtin/color_hsv.rs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    pub h: f32,
+    pub s: f32,
+    pub v: f32,
+    pub a: f32,
+}
+
+// Manual implementation because Hue minimal value (0.) is equal to maximal value (1.)
+impl PartialEq for ColorHsv {
+    fn eq(&self, other: &Self) -> bool {
+        (self.h == other.h || (self.h - other.h).abs() == 1.)
```
Comment (Ogeon) [inline godot-core/src/builtin/color_hsv.rs:90]: This assumes that the hue is normalized, right? It's not necessarily be the case with `h` being public. The same for `ApproxEq`.

**Comment by Ogeon** [comment]
> I believe that the integer-based algorithm could really allow for stable RGB<->HSV roundtrips, but there are unfortunately some strings attached that are imposed by RGB color being represented in Godot as floats. Maybe I will revisit the topic in the future, but for now I am pretty happy with the performance, ease of use and general usefullness of the float-based one. The integer-based one is more trickier to implement alongside the whole wrapper around it. You can look for the article referenced in original post for details, if it picks your interests :)

For sure, the integer based algorithm has its use as well. I didn't argue against the algorithm, itself. My list was meant as a "good to keep in mind" list and the specific algorithm still depends on the use case. Godot uses the same `Color` type for both linear and sRGB, so I wanted to raise the point about rounding as an example of when staying with floats matters for the visual outcome. :slightly_smiling_face: It could even be worth letting the user a choose the algorithm that fits them the best. Depends on how far you want to take it.

File: godot-core/src/builtin/color_hsv.rs (hunk 10 above)
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:90]: Good catch! I think I will remove the custom `PartialEq` implementation and will wrap the Hue on `ApproxEq` with suitable documentation

**Comment by StatisMike** [comment]
@Ogeon If a serious need arises, I will tackle this :) For clear separation at the beginning of the development of this feature, I included two different HSV structs: `ColorHsv` and `ColorHsvInt`, with each having their respective algorithms in `From<*>` implementations used. I believe if the topic should be revisited something like that should be used to reduce the conversions between integer <-> float, for both performance and precision reasons.

File: godot-core/src/builtin/color.rs (hunk 9 above)
Comment (Bromeon) [reply godot-core/src/builtin/color.rs:199]: We would then move a bit away from providing direct accessors (like Godot) and enforce conversion to `ColorHsv`.

But I think it's fair in terms of having one approach to do things, so I'm fine with removing the TODO. Could you document in 1 sentence of the `struct Color` docs, that HSV space can be accessed using `ColorHsv::from()`? Otherwise this is hard to discover.

File: godot-core/src/builtin/color_hsv.rs (hunk 10 above)
Comment (Bromeon) [reply godot-core/src/builtin/color_hsv.rs]: Please no implicit wraps -- make an `assert`. It is better that users are aware of potential logic errors.

File: godot-core/src/builtin/color_hsv.rs [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+/// Godot's [`Color`] built-in type supports mainly RGBA floating point-based notation. `ColorHsv` supports manipulating its HSVA
+/// representation by introducing conversion methods between itself and `Color`.
+///
+/// `ColorHsv` *is not* a [`GodotType`](crate::builtin::meta::GodotType). To use it in properties expecting `Color`, you need to convert
+/// it back to this type.
+///
+/// `Color` created by `Color::from(color_hsv)` is equal to one created by `Color::from_hsv()`, but the conversion time is approximately
+/// 4 times faster - partly because it avoids calls between Godot and Rust during conversion.
+///
+/// ## Conversions
```
[the remark this answers is not in the document]
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:22]: Maybe it would be better to have both of the methods in the `ColorHsv` namespace (as I guess we are keeping the `Color::from_hsv()` gateway to Godot builtin impl)?  `ColorHsv::from_rgba(Color) -> ColorHsv` and `ColorHsv::to_rgba(self) -> Color`

I feel like it will keep it cleaner: *if you are using methods from `ColorHsv`, you are using `gdext` custom implementation. To use Godot's builtin implementations using `gdext`, feel free to use the one on `Color`*

File: godot-core/src/builtin/color_hsv.rs [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+    pub s: f32,
+    pub v: f32,
+    pub a: f32,
+}
+
+impl ApproxEq for ColorHsv {
+    /// Approximate equality is checked with following conditions:
+    ///
+    /// - Hue values are wrapped before comparison, to make the logical sense first and foremost,
+    /// - All values are checked for equality with `10^-3` precision to compensate for potential data loss.
```
[the remark this answers is not in the document]
Comment (StatisMike) [reply godot-core/src/builtin/color_hsv.rs:109]: Thank you for the link, TBH I haven't discovered it :) Seems to be enough to compensate for the data loss.

Will also use methods from `FloatExt` in other relevant places

File: godot-core/src/builtin/color_hsv.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Bromeon) [reply godot-core/src/builtin/color_hsv.rs:22]: No, I think it's both more symmetric and more ergonomic to have two `to_*` methods that convert each type to the other.

What is maybe a bit weird is that then `TryFrom` lives outside the inherent impl, so there's an inconsistency between `rgb.to_hsv()` and `ColorHsv::try_from(&rgb)`. A `rgb.try_to_hsv()` might be nicer and would follow our "panic + `try_*` overload" pattern we apply in many APIs. I don't really know if genericity is useful here, plus `try_into()` on its own suffers from not having type inference.

**Comment by StatisMike** [comment]
@Bromeon Adressed the points and also added `serde` support for `ColorHsv` (of course feature-gated) + relevant tests

File: godot-core/src/builtin/color.rs [hunk 13]
```diff
@@ -294,6 +298,53 @@ impl Color {
         ]))
     }
 
+    /// Convert `Color` into [`ColorHsv`].
+    ///
+    /// ## Panics
+    ///
+    /// Method will panic if the RGBA values are outside of the valid range `0.0..=1.0`. You can use [`Color::normalized`] to ensure that
+    /// they are in range, or use [`Color::try_to_hsv`].
+    #[must_use]
```
Comment (Bromeon) [inline godot-core/src/builtin/color.rs:307]: We only use this attribute if there is potential ambiguity, e.g. in `Rect2::grow()`.

`to_*` methods are a well-known Rust convention for conversions, so it's very unlikely that someone would call this and ignore the return type. In fact there are even lints if you declare a `to_*` method that does something else.

File: godot-core/src/builtin/color_hsv.rs [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+/// HSVA floating-number Color representation.
+///
+/// Godot's [`Color`] built-in type supports mainly RGBA floating point-based notation. `ColorHsv` supports manipulating its HSVA
+/// representation by introducing conversion methods between itself and `Color`.
+///
+/// `ColorHsv` *is not* a [`GodotType`](crate::builtin::meta::GodotType). To use it in properties expecting `Color`, you need to convert
+/// it back to this type.
+///
+/// `Color` created by `Color::from(color_hsv)` is equal to one created by `Color::from_hsv(h, s, v)`, but the conversion time is
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:19]:
```suggestion
/// A `Color` created by [`ColorHsv::to_rgb()`] is equal to one created by `Color::from_hsv(h, s, v)`, but the conversion is
```

File: godot-core/src/builtin/color_hsv.rs [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    /// Convert `ColorHsv` into [`Color`].
+    ///
+    /// ## Panics
+    ///
+    /// Method will panic if the HSVA values are outside of the valid range `0.0..=1.0`. You can use [`ColorHsv::normalized_clamped_h`] or
+    /// [`ColorHsv::normalized_wrapped_h`] to ensure they are in range, or use [`ColorHsv::try_to_rgb`] implementation.
+    #[must_use]
+    pub fn to_rgb(self) -> Color {
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:182]: We can add the warning sign here (see explanation [here](https://godot-rust.github.io/docs/gdext/master/godot/#ergonomics-and-panics)). `Panics` should be H1.
```suggestion
    /// ⚠️ Convert `ColorHsv` into [`Color`].
    ///
    /// # Panics
    ///
    /// Method will panic if the HSVA values are outside of the valid range `0.0..=1.0`. You can use [`ColorHsv::normalized_clamped_h`] or
    /// [`ColorHsv::normalized_wrapped_h`] to ensure they are in range, or use [`ColorHsv::try_to_rgb`] implementation.
    pub fn to_rgb(self) -> Color {
```

File: godot-core/src/builtin/color_hsv.rs [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        match self.try_to_rgb() {
+            Ok(c) => c,
+            Err(e) => panic!("{}", e),
+        }
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:186]:
```suggestion
        self.try_to_rgb().unwrap_or_else(|e| panic!("{e}")),
```

File: godot-core/src/builtin/color_hsv.rs [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    }
+}
+
+impl Default for ColorHsv {
+    fn default() -> Self {
+        Self {
+            h: 0.,
+            s: 0.,
+            v: 0.,
+            a: 1.0,
```
Comment (Bromeon) [inline godot-core/src/builtin/color_hsv.rs:217]: Please use `0.0` here, not `0.`

File: itest/rust/src/builtin_tests/serde_test.rs [hunk 18]
```diff
@@ -9,7 +9,7 @@ use crate::framework::itest;
 use godot::builtin::{array, Array, GString, NodePath, StringName, Vector2i};
 use serde::{Deserialize, Serialize};
 
-fn serde_roundtrip<T>(value: &T, expected_json: &str)
+pub(crate) fn serde_roundtrip<T>(value: &T, expected_json: &str)
```
Comment (Bromeon) [inline itest/rust/src/builtin_tests/serde_test.rs:12]: The reason why this (and the surrounding module) was private is so that the checks are self-contained in `serde_test.rs` and we don't need additional `#[cfg]`s elsewhere.

Could you move the serde tests into this file and make the 2 symbols private again?

**Review by Bromeon — COMMENTED** [review]
Thanks!

**Review by Bromeon — APPROVED** [review]
Thanks a lot! 😊

**Comment by Bromeon** [comment]
How the hell did both this and #611 pass the merge queue?
The artifact name was wrong, a job failed and I cancelled both their CIs 🤦‍♂️
